firebase/firebase-tools
# Feature CORS-RFC1918 Support

### Description
Adding CORS-RFC1918 Support to Auth and Storage. 

Allows emulated firebase features to be accessed at localhost:#### from a site that is exposed via tunnel such as localtunnel or ngrok. See issue https://github.com/firebase/firebase-tools/issues/4227 & https://wicg.github.io/private-network-access/#headers

Change adds `access-control-allow-private-network=true` header to api responses when `access-control-request-private-network` is present. 

This change should be replaced in favor of a [expressjs/cors](https://github.com/expressjs/cors) option if adopted. See https://github.com/expressjs/cors/issues/236

If PR accepted please consider also adding  `access-control-allow-private-network=true` header to RTDB and Firestore.



### Scenarios Tested

Attempt to create a user/sign in, and upload file when site is exposed via a tunnel.
**Setup**
1. Setup a firebase project with hosting and emulators
2. Replace index.html in public folder with below
3. Emulate project 
4. Expose site via tunnel using [localtunnel/localtunnel](https://github.com/localtunnel/localtunnel) `lt -p 5000 -s firebase-emulator-cors-test`
5. Navigate to https://firebase-emulator-cors-test.loca.lt (or url output by localtunnel)

**Test Authentication**

1. Click "Create User"
2. Verify user was created (see console & emulator ui)
3. Click "Authenticate User"
4. Verify user was authenticated (see console)

**Test Storage**
1. Click Choose files
2. Verify file was uploaded (see console & emulator ui)



```
<!DOCTYPE html>
<html lang="en">
<head>
  <title>App</title>
</head>
<body>
  
  <body>
    <div>
      <button id="createUser">Create User</button>
      <button id="authenticateUser">Authenticate User</button>
      <input type="file" id="files" name="files[]" multiple />
      
    </div>
    <script type="module">
      
      import { initializeApp } from 'https://www.gstatic.com/firebasejs/9.6.7/firebase-app.js'
    
      import { 
        getAuth, 
        connectAuthEmulator, 
        signInWithEmailAndPassword,
        createUserWithEmailAndPassword,
      } from 'https://www.gstatic.com/firebasejs/9.6.7/firebase-auth.js'
      
      import { 
        getStorage, 
        ref, 
        uploadBytesResumable, 
        getDownloadURL,
        connectStorageEmulator
      } from "https://www.gstatic.com/firebasejs/9.6.7/firebase-storage.js";


      const email = "name@fake.com"
      const password = "password"

      const firebaseConfig = {
        apiKey: "AIzaSyDrzYTAKem9ut-r_iV-vIDZggnsjvttfQM",
        authDomain: "fir-emulator-test-40bf3.firebaseapp.com",
        projectId: "fir-emulator-test-40bf3",
        storageBucket: "fir-emulator-test-40bf3.appspot.com",
        messagingSenderId: "1033607461892",
        appId: "1:1033607461892:web:4bfee036d7c11e4f7a4951"
      };

      const app = initializeApp(firebaseConfig);
      const auth = getAuth()
      const storage = getStorage();

      connectAuthEmulator(auth, "http://localhost:9099" );
      connectStorageEmulator(storage, "localhost", 9199)
        
      async function signOut () {
        console.log("Signing out");
        await auth.signOut()
        console.log("Signed out");
      }
      async function authenticateUser () {
        await signOut()
        console.log("Signing in");
        const credential = await signInWithEmailAndPassword(auth, email, password)
        console.log("Signed in", credential)
      }
      async function createUser () {
        console.log('Creating User');
        const credential = await createUserWithEmailAndPassword(auth, email, password)
        console.log("User created", credential)
      }

      auth.onAuthStateChanged(console.log)

      function uploadFile (e) {
        console.log(e);
        const file = e.srcElement.files[0];

        const storageRef = ref(storage, `files/${file.name}`);
        const uploadTask = uploadBytesResumable(storageRef, file);

        uploadTask.on('state_changed', 
          (snapshot) => {
            const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
            console.log("State: ", snapshot.state)
            console.log('Upload is ' + progress + '% done');
          }, 
          (error) => {
            // Handle unsuccessful uploads
          }, 
          () => {
            getDownloadURL(uploadTask.snapshot.ref).then((downloadURL) => {
              console.log('File available at', downloadURL);
            });
          }
        );
      }

      
      document.getElementById('createUser').addEventListener('click', createUser, true);
      document.getElementById('authenticateUser').addEventListener('click', authenticateUser, true);
      document.getElementById('files').addEventListener('change', uploadFile, true);
    </script>
  </body>
</body>
</html>
```


### Sample Commands



N/A
